packages/firebase_performance/lib/src/trace.dart

```diff
@@ -0,0 +1,155 @@
+      'id': _id,
+      'name': _name,
+    });
+  }
+
+  /// Stops this trace.
+  ///
+  /// Using 'await' with this method is only necessary when accurate timing
+  /// is relevant.
+  Future<void> stop() {
+    assert(!_hasStopped,
+        "Trace '$_name' has already stopped, should not stop again!");
+    assert(
+        _hasStarted, "Trace '$_name' has not been started so unable to stop!");
+
+    final Map<String, dynamic> data = <String, dynamic>{
+      'id': _id,
+      'name': _name,
+      'counters': _counters,
+      'attributes': _attributes,
+    };
+
+    _hasStopped = true;
+    return FirebasePerformance.channel.invokeMethod('Trace#stop', data);
+  }
+
+  /// Increments the counter in this trace with the given [name] by given value.
+  ///
+  /// Increments the counter in this trace with the given name by given value.
+  /// If a counter does not already exist, a new one will be created. If the
+  /// trace has not been started or has already been stopped, returns
+  /// immediately without taking action.
+  ///
+  /// [name]: Name of the counter to be incremented. Requires no leading or
+  /// trailing whitespace, no leading underscore _ character, max length of
+  /// [maxCounterKeyLength] characters.
+  ///
+  /// [incrementBy]: Amount by which the counter has to be incremented.
+  void incrementCounter(String name, [int incrementBy = 1]) {
+    assert(!_hasStopped,
+        "Connot increment counter $name. Trace '$_name' has already stopped!");
+    assert(name != null,
+        "Cannot increment counter. Counter name is invalid. (Counter name must not be null)");
+    assert(!name.startsWith(new RegExp(r'[_\s]')),
+        "Cannot increment counter $name. Counter name is invalid. (Counter name must not start with '_' or space)");
+    assert(!name.contains(new RegExp(r'[_\s]$')),
+        "Cannot increment counter $name. Counter name is invalid. (Counter name must not end with '_' or space)");
+    assert(name.length <= 32,
+        "Cannot increment counter $name. Counter name is invalid. (Counter name must not exceed 32 characters)");
+
+    _counters.putIfAbsent(name, () => 0);
+    _counters[name] += incrementBy;
+  }
+
+  /// Sets a String [value] for the specified [attribute].
+  ///
+  /// Sets a String value for the specified attribute. Updates the value of the
+  /// attribute if the attribute already exists. If the trace has been stopped,
+  /// this method returns without adding the attribute. The maximum number of
+  /// attributes that can be added to a Trace are [maxTraceCustomAttributes].
+  ///
+  /// [attribute]: Name of the attribute. Max length of [maxAttributeKeyLength]
+  /// characters.
+  ///
+  /// [value]: Value of the attribute. Max length of [maxAttributeValueLength]
+  /// characters.
+  void putAttribute(String attribute, String value) {
+    assert(!_hasStopped,
+        "Can not set attriubte $attribute. Trace '$_name' has already stopped!");
```

Comment:
Typo here. I am leaning towards just removing the assertion descriptions, I don't think they add much clarity and we don't use them anywhere else in this repo. If we do want descriptions, consider throwing a `FlutterError`, see the flutter/flutter repo for [examples](https://github.com/flutter/flutter/blob/cc1cf9e12a226ec67875a857c2fcc097541781b1/packages/flutter/lib/src/widgets/table.dart#L110).